Automattic/wp-calypso
# Orders: Update state infrastructure to allow sending a query to fetch orders

This (big 😩 ) PR updates the orders state handling in `sites` & `ui` so that we can fetch and track orders based on queries, not just page numbers. The query will _probably_ only ever be a page/status combo, but could be expanded on later.

- Update `fetchOrders` to accept a query, it defaults to page 1 and all order statuses
- Update selectors `areOrdersLoading`, `areOrdersLoaded` to accept queries
- Fetch orders from the API using the 
- Add a currentStatus reducer, `setCurrentQuery` action, and `getOrdersCurrentStatus` to the UI state subtree to keep track of currently-viewed status, like we already track page
- Make sure the actions, reducers, and selectors know how to handle serialized queries as keys for order mapping, and the totals list
- Change `getTotalOrdersPages` to `getTotalOrders`, since we don't need to track pages
- Add pagination to orders list, which now uses the new query-fetching

**To test:**

- Run the test suite: `npm test`
- Run calypso and look at the places we use orders
  - The orders list
  - The dashboard new order display, for sites with orders
  - The dashboard "no orders view", for sites with no orders

Note: Eventually we'll get new orders directly from the API, rather than fetching all orders and filtering, but that needs an API update first, see https://github.com/woocommerce/wc-api-dev/issues/26
